packages/jupyter-ai/src/components/rendermime-markdown.tsx

```diff
@@ -39,34 +43,52 @@ function RendermimeMarkdownBase(props: RendermimeMarkdownProps): JSX.Element {
       const renderer = props.rmRegistry.createRenderer(MD_MIME_TYPE);
       await renderer.renderModel(model);
       props.rmRegistry.latexTypesetter?.typeset(renderer.node);
-
-      // Attach CopyButton to each <pre> block
-      if (containerRef.current && renderer.node) {
-        const preBlocks = renderer.node.querySelectorAll('pre');
-        preBlocks.forEach(preBlock => {
-          const copyButtonContainer = document.createElement('div');
-          preBlock.parentNode?.insertBefore(
-            copyButtonContainer,
-            preBlock.nextSibling
-          );
-          ReactDOM.render(
-            <CopyButton value={preBlock.textContent || ''} />,
-            copyButtonContainer
-          );
-        });
+      if (!renderer.node) {
+        throw new Error(
+          'Rendermime was unable to render Markdown content within a chat message. Please report this upstream to Jupyter AI on GitHub.'
+        );
       }
 
+      const newCodeToolbarDefns: [HTMLDivElement, CodeToolbarProps][] = [];
+
+      // Attach CodeToolbar root element to each <pre> block
+      const preBlocks = renderer.node.querySelectorAll('pre');
```

Comment:
I see this even on `pre` blocks that indicate an error in contacting the language model. Is it possible to avoid adding the CodeToolbar in this case?

Reply:
Not at this time. The error message is returned directly from the backend as a chat message, so there's no way to distinguish code blocks in error messages from code blocks in "normal" messages reliably. This can be done in a follow-up PR.

I don't agree the code toolbar should be entirely hidden in this case, since a user may still want to copy the error message to open a GitHub issue. But it probably makes sense to hide the "Replace" and "Insert" buttons.